### DISCUSSION TITLE
SQL Database view and query console

### DISCUSSION THREAD

**Discussion by torikki-tou** [zed-industries/zed#24000] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

Two main reasons i still mainly use Jetbrains IDEs instead of Zed is git tooling and integration with databases. There is issue for git but i didn't found one about databases.

So basically in IDEs like PyCharm and GoLand you can connect to almost any datasource to observe and query data. Would be nice to have it in Zed. Adding some of the most popular relational and NoSQL DBs (PostgreSQL, SQLite, MySQL, Redis, MongoDB) should be the first step in my opinion and then an interface for integration can be provided.

### Environment

Zed: v0.163.2 (Zed)
OS: macOS 15.1.1
Memory: 32 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help present your vision of the feature

<details><summary>screenshots</summary>

![image](https://github.com/user-attachments/assets/10f6191d-d919-4902-b884-b5b9f9cda026)
![image](https://github.com/user-attachments/assets/f90d7b7a-8edd-4fb3-a052-1f707b98815b)
![image](https://github.com/user-attachments/assets/a8e1a7bb-e0d4-4176-a344-7a11f723b9a4)

</details>

**Comment by notpeter** [comment]
Practically speaking JetBrains has a big leg up here, not just because they have head start (DataGrip turns [9 years old in a week](https://blog.jetbrains.com/datagrip/2022/12/16/datagrip-turns-7/)) but because Jetbrains IDEs run on the JRE (Java/Kotlin).  Nearly every database ships a JDBC driver with a relatively well defined interface which so you can support obscure databases (CockroachDB, DB2, SQL Server, SAP Hana) on any platform as long as cross-platform JDBC-driver available (even if those drivers are closed source).  This would a harder hill to climb if it were written in Rust like the rest of Zed.

Thanks for reporting.

**Comment by indrajit-roy** [comment]
How about using libs like [`rusqlite`](https://github.com/rusqlite/rusqlite)? Since the bindings are already created, it should be relatively simpler? I just can't identify the APIs I can use to create an extension like VSCode's [sqlite3-editor](https://github.com/yy0931/sqlite3-editor) where I can just select the `.db` file and it gives me a UI to interact with my database.

**Comment by minibikini** [comment]
It'd be so cool to have a simple Postgres GUI with autocomplete and an AI assistant built in.

**Comment by Mnior** [comment]
> drivers

This is only one side of the coin. And it seems wrong to me to consider as the only option that the Zed team should support these drivers.

The problem is completely different - few editors can boast TableView in text mode. And it is the absence of this type of text representation that is the stumbling block for integrating such things.
It is not necessary that the work with data is carried out on the Zed side, let LangugeServer or something similar do it, the main thing is a convenient tabular representation of data.

And this is not only working with SQL queries, but also trivially opening CSV files and any similar table formats.
And if you look at how they usually try to implement this - as formatting, then the error in the approach is immediately visible. But it is necessary that the horizontal display (columns) be virtualized in the same way as it is done now, for example, for the function header - vertical virtualization of rows.

But there are many ways to display tabular data, and each SQL IDE does it in its own way (extended windows, tabs, inline, cached, post-processing), not to mention the methods of working with connections, etc. And with TextTableView as a presentation mechanism, it will open a new page in the history of building a customizable environment for working with data.

**Reply by torikki-tou** [reply]
I think you right! What i mainly want from Zed team is such infrastructure like tabular views. Drivers will be handled by community in some way.

**Comment by ghost** [comment]
up vote for this, love jetbrains ides but zed is great alternative.

**Reply by maxvanceffer** [reply]
Zed is best even now.

**Comment by davidjb99** [comment]
It would be good to get this baked into Zed, it’s a popular feature in PyCharm. At the very least enabling extensions to do this would be a good start. VScode has an extension which is good enough, I don’t think it’s possible to build such an extension in Zed yet..

https://marketplace.visualstudio.com/items?itemName=cweijan.vscode-database-client2

**Comment by tshazar** [comment]
First class multi dialect sql support with AI data modeling and sql code creation coupled with the existing agentic editing would be next level

**Comment by lukejans** [comment]
VScode offers a similar UI for viewing tables, writing & running queries, and auto-completion for table columns and what not. 
> this is all fake data for school !!!

<img width="1582" height="1034" alt="Screenshot 2025-09-22 at 5 56 14 PM" src="https://github.com/user-attachments/assets/bbf5992a-09d2-49ec-9fdd-35b786c6169d" />

<img width="1582" height="1034" alt="Screenshot 2025-09-22 at 6 01 50 PM" src="https://github.com/user-attachments/assets/d12fb753-3be2-4882-be34-a7cb1ba80a1c" />

<img width="1582" height="1034" alt="Screenshot 2025-09-22 at 6 02 05 PM" src="https://github.com/user-attachments/assets/04102a4c-42aa-4ead-8c57-108783fa83d5" />

**Comment by HalavicH** [comment]
Hey folks 👋

I’ve been exploring this space recently, and as a first step toward better tabular data support in Zed, I opened a draft PR that adds a **CSV/TSV preview with a real table UI** (resizable columns, sorting, cell selection, quote-aware parsing, etc.):

**PR:** [https://github.com/zed-industries/zed/pull/44344](https://github.com/zed-industries/zed/pull/44344)

It’s still very much a POC, but several people in this thread mentioned wanting a generic “table viewer” or a lightweight foundation for DB query results, so I thought it might be relevant here. My long-term goal is also to move toward database-client features, and CSV preview felt like a good first step.

Totally open to feedback, ideas, or any advice if anyone wants to chime in. 🙌

**Reply by torikki-tou** [reply]
Great job! I am very proud of being the topic starter of such a thing. I wish I could write rust code and could help you out

**Reply by mjovanc** [reply]
Great job. This is something that would take Zed to the next level.

**Reply by ishazel** [reply]
That's awesome! Waiting for the day I can fully ditch VS Code 😊

**Comment by ronannnn** [comment]
up vote for this

**Comment by yangchnet** [comment]
Perhaps we could take inspiration from [dbx](https://github.com/t8y2/dbx), which is a good project, but I prefer zed's 120 fps.

**Comment by Pioneer-1-1** [comment]
A lot of this looks like it could also be useful for tabular data previous i.e. something like how vscode lets you inspect the contents of numpy arrays
